core/runtime_modules.rs

```diff
@@ -0,0 +1,206 @@
+  }
+
+  pub fn with_ops(
+    js_files: Vec<SourcePair>,
+    ops: Vec<OpPair>,
+    opstate_fn: Option<Box<OpStateFn>>,
+  ) -> Self {
+    Self::new(Some(js_files), Some(ops), opstate_fn)
+  }
+}
+
+impl JsRuntimeModule for BasicModule {
+  fn init_js(&self) -> Result<Vec<SourcePair>, AnyError> {
+    Ok(match &self.js_files {
+      Some(files) => files.clone(),
+      None => vec![],
+    })
+  }
+
+  fn init_ops(&mut self, registrar: RcOpRegistrar) -> Result<(), AnyError> {
+    // NOTE: not idempotent
+    // TODO: fail if called twice ?
+    if let Some(ops) = self.ops.take() {
+      for (name, opfn) in ops {
+        registrar.borrow_mut().register_op(name, opfn);
+      }
+    }
+    Ok(())
+  }
+
+  fn init_state(&self, state: &mut OpState) -> Result<(), AnyError> {
+    match &self.opstate_fn {
+      Some(ofn) => ofn(state),
+      None => Ok(()),
+    }
+  }
+}
+
+// MultiModule allows grouping multiple sub-JsRuntimeModules into one,
+// allowing things such as:
+// ```
+// fn web_modules(args: WebModuleArgs) -> MultiModule {
+//  MultiModule::new(vec![deno_url::init(), deno_fetch::init(...), ...])
+// }
+// ```
+pub struct MultiModule<'s> {
+  pub modules: Vec<Box<dyn JsRuntimeModule + 's>>,
+}
+
+impl MultiModule<'_> {
+  fn new<'s>(modules: &mut Vec<impl JsRuntimeModule + 's>) -> MultiModule<'s> {
+    let modules = modules
+      .drain(..)
+      .map(|m| Box::<dyn JsRuntimeModule + 's>::from(Box::new(m)))
+      .collect();
+    MultiModule { modules }
+  }
+}
+
+impl JsRuntimeModule for MultiModule<'_> {
+  fn init_js(&self) -> Result<Vec<SourcePair>, AnyError> {
+    Ok(
+      self
+        .modules
+        .iter()
+        .map(|m| m.init_js().unwrap())
+        .flatten()
+        .collect(),
+    )
+  }
+
+  fn init_ops(&mut self, registrar: RcOpRegistrar) -> Result<(), AnyError> {
+    for m in self.modules.iter_mut() {
+      m.init_ops(registrar.clone())?;
+    }
+    Ok(())
+  }
+}
+
+// The OpRegistrar trait allows building op "middleware" such as:
+// OpMetrics, OpTracing or OpDisabler that wrap OpFns for profiling, debugging, etc...
+// JsRuntime is itself an OpRegistrar
+pub trait OpRegistrar {
+  fn register_op(&mut self, name: &str, op_fn: Box<OpFn>) -> OpId;
+  // register_minimal_op_sync(...)
+  // register_minimal_op_async(...)
+  // register_json_op_sync(...)
+  // register_json_op_async(...)
+}
+
+////
+// Helper macros to reduce verbosity / redundant decls
+////
+#[macro_export]
+macro_rules! include_js_files {
+  (root $root:literal, $($file:literal,)+) => {
+    vec![
+      $((
+        concat!($root, "/", $file),
+        include_str!($file),
+      ),)+
+    ]
+  };
+}
+
+#[macro_export]
+macro_rules! declare_ops {
+  (with($wrapper:path), $($path:ident::$opfn:ident,)+) => {
+    vec![$((
+      stringify!($opfn),
+      $wrapper($path::$opfn),
+    ),)+]
+  };
+
+  (with($wrapper:path), $($opfn:ident,)+) => {
+    vec![$((
+      stringify!($opfn),
+      $wrapper($opfn),
+    ),)+]
+  };
+}
```

Comment:
I'm a bit wary of introducing macros. Although it reduces code, I worry that it obfuscates what's happening. That said, these seem relatively simple... Maybe consider doing these in a follow up PR.

Reply:
I agree, I'm a bit wary of the `with()` middleware wrapper. This PR already introduces quite a few new concepts, I suggest to skip middleware implementation for a follow up PR.

Reply:
@ry @bartlomieju What about something like:

```rust
    declare_ops!(json_op_async[
      op_webgpu_request_adapter,
      op_webgpu_request_device,
    ]),
```

That's a little easier to read, it conveys that `json_op_async` wraps `[...]` and hints that it returns a `Vec`

Reply:
It could also be:

```rust
    declare_ops!(json_op_async, [
      op_webgpu_request_adapter,
      op_webgpu_request_device,
    ]),
```

If we want it to look more like a map expression (notice `, [`)